[Bug] sysbench1000w read_write_100 hangs with lockservice missing range last wait key

## Description

In nightly regression on TKE, `sysbench1000w.read_write_100` hung until the GitHub Actions job timeout. The sysbench process did not finish its configured `--time=300`; instead TPS/QPS dropped to zero and remained there until the 280-minute job timeout killed the process.

Job: https://github.com/matrixorigin/mo-nightly-regression/actions/runs/26458903346/job/77948823011

## Environment

- Workflow: `MO Main Nightly Regression On TKE`
- Job: `Main Nightly Regression Test / SYSBENCH OLTP TEST 1000W`
- Namespace: `mo-main-commit-a21829834-20260526`
- Cluster: `nightly-regression-dis`
- Commit: `commit-a21829834`
- Workload: `sysbench1000w.read_write_100`

Sysbench command:

```bash
sysbench --mysql-host=172.16.59.245 --mysql-port=6001 --mysql-user=dump --mysql-password=*** --db-ps-mode=disable --mysql-ignore-errors=20701 oltp_read_write.lua --mysql-db=sysbench_db --tables=10 --table_size=10000000 --threads=100 --time=300 --report-interval=10 --point_selects=1 run
```

## Timeline

All times below are Beijing time.

- `2026-05-27 04:57:30`: `read_write_100` started.
- First ~100 seconds were normal, TPS/QPS were non-zero.
- `2026-05-27 04:59:21`: sysbench started reporting `tps: 0.00`, `qps: 0.00`, `err/s: 0.00`.
- TPS/QPS stayed at zero until the job was cancelled by GitHub Actions timeout.
- `2026-05-27 08:49:35`: job hit `timeout-minutes: 280` and was cancelled.

Example sysbench behavior:

```text
[100s] thds: 100 tps: 2322.60 qps: ...
[110s] thds: 100 tps: 0.00 qps: 0.00 err/s: 0.00 reconn/s: 0.00
...
[13920s] thds: 100 tps: 0.00 qps: 0.00 err/s: 0.00 reconn/s: 0.00
##[error]The operation was canceled.
```

## Loki findings

Around `2026-05-27 04:59:10`, shortly before TPS/QPS dropped to zero, multiple TP-CN pods logged row-lock upgrade messages on `sbtest10`:

```text
Row lock upgraded to table lock for SQL : UPDATE sbtest10 SET k=k+1 WHERE id=...
Row lock upgraded to table lock for SQL : DELETE FROM sbtest10 WHERE id=...
```

These appeared on all three TP-CN pods:

```text
nightly-regression-dis-tp-cn-c2kck
nightly-regression-dis-tp-cn-cmjf9
nightly-regression-dis-tp-cn-t4m84
```

In the same time window, `nightly-regression-dis-tp-cn-cmjf9` logged lockservice errors:

```text
level=ERROR
name=cn-service.lockservice
caller=lockservice/lock_table_local.go:551
msg="missing range last wait key when moving waiter to next conflict"
table=272842
```

Example log:

```text
{"level":"ERROR","time":"2026/05/26 20:59:10.602842 +0000","name":"cn-service.lockservice","caller":"lockservice/lock_table_local.go:551","msg":"missing range last wait key when moving waiter to next conflict","table":272842,"last-wait-key":"OhB/////","next-conflict-key":"OhcfloU="}
```

No obvious `panic`, `OOM`, `restart`, `crash`, or `SIGKILL` evidence was found in the same interval.

There were also many accompanying errors such as:

```text
dedup load did not complete
i/o timeout
```

But the lockservice error and row-lock/table-lock messages line up most closely with the moment sysbench stopped making progress.

## Pyroscope memory profile

The memory profile does not show a continuous increase to OOM. It shows a short memory peak around the same time as the lockservice issue, then memory drops later.

Overall `memory:inuse_space`:

```text
start nearby: ~32.87 GiB
peak:         ~38.85 GiB at 2026-05-27 04:59:45
later low:    ~20.36 GiB
timeout near: ~22.79 GiB
```

Pod-level peak around the incident:

```text
nightly-regression-dis-dn-0          max ~11.64 GiB at 04:59:45
nightly-regression-dis-tp-cn-c2kck   max ~11.08 GiB at 05:00:45
nightly-regression-dis-tp-cn-cmjf9   max ~ 7.76 GiB at 05:02:45
nightly-regression-dis-tp-cn-t4m84   max ~ 8.11 GiB at 04:59:30
```

Top memory profile hotspots include:

```text
objectio.LoadObjectMetaByExtent
objectio.ReadBloomFilter
io.ReadAll
container/vector.NewVecFromReuse
logtailreplay.PartitionState.HandleRowsInsert
logtailreplay.PartitionState.HandleRowsDelete
```

This looks like accompanying objectio/logtail memory pressure, but not the direct reason for the 280-minute hang.

## Expected behavior

`sysbench1000w.read_write_100` should finish after `--time=300` and output the final sysbench summary.

## Actual behavior

TPS/QPS dropped to zero after about 110 seconds. Sysbench did not exit after `--time=300`, and the GitHub Actions job was eventually cancelled by the 280-minute timeout.

## Current analysis

This looks like an MO-side lockservice/transaction lock waiting hang triggered by the `read_write_100` workload, rather than a test runner or GitHub Actions timeout issue. The most suspicious log is:

```text
lockservice/lock_table_local.go:551
missing range last wait key when moving waiter to next conflict
```

The SQLs around the issue are mostly `UPDATE`/`DELETE` on `sbtest10`, and row locks were upgraded to table locks immediately before sysbench stopped making progress.

--- a/pkg/lockservice/lock.go
+++ b/pkg/lockservice/lock.go
@@ -210,15 +210,15 @@ func (l Lock) closeWaiter(w *waiter, logger *log.MOLogger) bool {
 	return canRemove
 }
 
-func (l Lock) removeWaiter(w *waiter, logger *log.MOLogger) bool {
+func (l Lock) removeWaiter(w *waiter, logger *log.MOLogger) (bool, bool) {
 	removed, wasFirst := l.waiters.remove(w)
 	if !removed {
-		return l.isEmpty()
+		return false, l.isEmpty()
 	}
 	if l.holders.size() == 0 && wasFirst && l.waiters.size() > 0 {
 		l.waiters.notify(notifyValue{defChanged: l.isLockTableDefChanged()})
 	}
-	return l.isEmpty()
+	return true, l.isEmpty()
 }
 
 func (l Lock) closeTxn(

--- a/pkg/lockservice/lock.go
+++ b/pkg/lockservice/lock.go
@@ -210,6 +210,17 @@ func (l Lock) closeWaiter(w *waiter, logger *log.MOLogger) bool {
 	return canRemove
 }
 
+func (l Lock) removeWaiter(w *waiter, logger *log.MOLogger) bool {
+	removed, wasFirst := l.waiters.remove(w)
+	if !removed {
+		return l.isEmpty()
+	}
+	if l.holders.size() == 0 && wasFirst && l.waiters.size() > 0 {
+		l.waiters.notify(notifyValue{defChanged: l.isLockTableDefChanged()})
+	}
+	return l.isEmpty()
+}
+
 func (l Lock) closeTxn(
 	txn *activeTxn,
 	notify notifyValue) (lockCanRemoved bool) {

--- a/pkg/lockservice/lock_table_local.go
+++ b/pkg/lockservice/lock_table_local.go
@@ -515,6 +515,10 @@ func (l *localLockTable) handleLockConflictLocked(
 		return ErrLockConflict
 	}
 
+	if c.opts.Granularity == pb.Granularity_Range {
+		l.closeRangeLastWaiterLocked(c)
+	}
+
 	c.w.conflictKey.Store(&key)
 	c.w.lt.Store(l)
 	c.w.waitFor = c.w.waitFor[:0]
@@ -541,25 +545,40 @@ func (l *localLockTable) handleLockConflictLocked(
 	// waiter added, we need to active deadlock check.
 	logLocalLockWaitOn(l.logger, c.txn, l.bind.Table, c.w, key, conflictWith)
 
-	if c.opts.Granularity != pb.Granularity_Range {
-		return nil
-	}
-
-	if len(c.rangeLastWaitKey) > 0 {
-		v, ok := l.mu.store.Get(c.rangeLastWaitKey)
-		if !ok {
-			l.logger.Error("missing range last wait key when moving waiter to next conflict",
-				zap.Uint64("table", l.bind.Table),
-				zap.String("txn", c.txn.txnKey),
-				zap.Binary("last-wait-key", c.rangeLastWaitKey),
-				zap.Binary("next-conflict-key", key))
-		}
-		if ok && v.closeWaiter(c.w, l.logger) {
-			l.mu.store.Delete(c.rangeLastWaitKey)
-		}
-	}
 	c.rangeLastWaitKey = key
 	return nil
+}
+
+func (l *localLockTable) closeRangeLastWaiterLocked(c *lockContext) {
+	if len(c.rangeLastWaitKey) == 0 {
+		return
+	}
+
+	v, ok := l.mu.store.Get(c.rangeLastWaitKey)
+	if ok {
+		if v.removeWaiter(c.w, l.logger) {
+			l.mu.store.Delete(c.rangeLastWaitKey)
+		}
+		c.rangeLastWaitKey = nil
+		return
+	}
+
+	l.logger.Error("missing range last wait key when moving waiter to next conflict",
+		zap.Uint64("table", l.bind.Table),
+		zap.String("txn", c.txn.txnKey),
+		zap.Binary("last-wait-key", c.rangeLastWaitKey))
+
+	var deleteKeys [][]byte
+	l.mu.store.Iter(func(key []byte, lock Lock) bool {
+		if lock.removeWaiter(c.w, l.logger) {
+			deleteKeys = append(deleteKeys, append([]byte(nil), key...))
+		}
+		return true
+	})
+	for _, key := range deleteKeys {
+		l.mu.store.Delete(key)
+	}
+	c.rangeLastWaitKey = nil
 }
 
 func (l *localLockTable) addRangeLockLocked(

--- a/pkg/lockservice/lock_table_local.go
+++ b/pkg/lockservice/lock_table_local.go
@@ -556,21 +556,29 @@ func (l *localLockTable) closeRangeLastWaiterLocked(c *lockContext) {
 
 	v, ok := l.mu.store.Get(c.rangeLastWaitKey)
 	if ok {
-		if v.removeWaiter(c.w, l.logger) {
+		removed, empty := v.removeWaiter(c.w, l.logger)
+		if removed {
+			if empty {
+				l.mu.store.Delete(c.rangeLastWaitKey)
+			}
+			c.rangeLastWaitKey = nil
+			return
+		}
+		if empty {
 			l.mu.store.Delete(c.rangeLastWaitKey)
 		}
-		c.rangeLastWaitKey = nil
-		return
 	}
 
 	l.logger.Error("missing range last wait key when moving waiter to next conflict",
 		zap.Uint64("table", l.bind.Table),
 		zap.String("txn", c.txn.txnKey),
-		zap.Binary("last-wait-key", c.rangeLastWaitKey))
+		zap.Binary("last-wait-key", c.rangeLastWaitKey),
+		zap.Bool("last-wait-key-exists", ok))
 
 	var deleteKeys [][]byte
 	l.mu.store.Iter(func(key []byte, lock Lock) bool {
-		if lock.removeWaiter(c.w, l.logger) {
+		removed, empty := lock.removeWaiter(c.w, l.logger)
+		if removed && empty {
 			deleteKeys = append(deleteKeys, append([]byte(nil), key...))
 		}
 		return true

--- a/pkg/lockservice/lock_table_local_test.go
+++ b/pkg/lockservice/lock_table_local_test.go
@@ -1716,6 +1716,17 @@ func TestHandleLockConflictLockedLogOnMissingRangeKey(t *testing.T) {
 				waitTxn := pb.WaitTxn{TxnID: txnID, CreatedOn: "test"}
 				w := acquireWaiter(waitTxn, "test", logger)
 
+				staleKey := []byte{2}
+				staleWaiters := newWaiterQueue()
+				staleWaiters.init(logger)
+				staleWaiters.put(w)
+				lt.mu.store.Add(staleKey, Lock{
+					createAt: time.Now(),
+					holders:  newHolders(),
+					waiters:  staleWaiters,
+				})
+
+				nextConflictKey := []byte{1}
 				holderTxnID := []byte("holder1")
 				holderWaitTxn := pb.WaitTxn{TxnID: holderTxnID, CreatedOn: "test"}
 				h := newHolders()
@@ -1727,6 +1738,7 @@ func TestHandleLockConflictLockedLogOnMissingRangeKey(t *testing.T) {
 					holders:  h,
 					waiters:  wq,
 				}
+				lt.mu.store.Add(nextConflictKey, conflictWith)
 
 				c := &lockContext{
 					ctx:     context.Background(),
@@ -1744,11 +1756,18 @@ func TestHandleLockConflictLockedLogOnMissingRangeKey(t *testing.T) {
 					result:           pb.Result{},
 				}
 
-				// rangeLastWaitKey {99} is not in the store, so the error log
-				// branch (previously a panic) should execute without crashing.
-				err := lt.handleLockConflictLocked(c, []byte{1}, conflictWith)
+				// rangeLastWaitKey {99} is not in the store, but the same waiter
+				// may have been moved to another queue by a range merge. Leaving it
+				// in that stale no-holder lock would block later lockers forever.
+				err := lt.handleLockConflictLocked(c, nextConflictKey, conflictWith)
 				assert.NoError(t, err)
-				assert.Equal(t, []byte{1}, c.rangeLastWaitKey)
+				assert.Equal(t, nextConflictKey, c.rangeLastWaitKey)
+				_, ok := lt.mu.store.Get(staleKey)
+				assert.False(t, ok)
+
+				nextLock, ok := lt.mu.store.Get(nextConflictKey)
+				require.True(t, ok)
+				assert.Equal(t, 1, nextLock.waiters.size())
 			})
 		},
 	)

--- a/pkg/lockservice/lock_table_local_test.go
+++ b/pkg/lockservice/lock_table_local_test.go
@@ -1726,6 +1726,17 @@ func TestHandleLockConflictLockedLogOnMissingRangeKey(t *testing.T) {
 					waiters:  staleWaiters,
 				})
 
+				recreatedKey := []byte{99}
+				recreatedHolders := newHolders()
+				recreatedHolders.add(pb.WaitTxn{TxnID: []byte("recreated-holder"), CreatedOn: "test"})
+				recreatedWaiters := newWaiterQueue()
+				recreatedWaiters.init(logger)
+				lt.mu.store.Add(recreatedKey, Lock{
+					createAt: time.Now(),
+					holders:  recreatedHolders,
+					waiters:  recreatedWaiters,
+				})
+
 				nextConflictKey := []byte{1}
 				holderTxnID := []byte("holder1")
 				holderWaitTxn := pb.WaitTxn{TxnID: holderTxnID, CreatedOn: "test"}
@@ -1752,18 +1763,20 @@ func TestHandleLockConflictLockedLogOnMissingRangeKey(t *testing.T) {
 						},
 					},
 					w:                w,
-					rangeLastWaitKey: []byte{99},
+					rangeLastWaitKey: recreatedKey,
 					result:           pb.Result{},
 				}
 
-				// rangeLastWaitKey {99} is not in the store, but the same waiter
-				// may have been moved to another queue by a range merge. Leaving it
-				// in that stale no-holder lock would block later lockers forever.
+				// rangeLastWaitKey was removed by a range merge and recreated by
+				// another txn without this waiter, but the waiter may still exist
+				// in another stale no-holder lock queue.
 				err := lt.handleLockConflictLocked(c, nextConflictKey, conflictWith)
 				assert.NoError(t, err)
 				assert.Equal(t, nextConflictKey, c.rangeLastWaitKey)
 				_, ok := lt.mu.store.Get(staleKey)
 				assert.False(t, ok)
+				_, ok = lt.mu.store.Get(recreatedKey)
+				assert.True(t, ok)
 
 				nextLock, ok := lt.mu.store.Get(nextConflictKey)
 				require.True(t, ok)

--- a/pkg/lockservice/waiter_queue.go
+++ b/pkg/lockservice/waiter_queue.go
@@ -31,6 +31,7 @@ type waiterQueue interface {
 	resetCommittedAt(timestamp.Timestamp)
 	moveTo(to waiterQueue)
 	put(...*waiter)
+	remove(*waiter) (bool, bool)
 	notify(value notifyValue)
 	notifyAll(value notifyValue)
 	first() *waiter
@@ -77,6 +78,28 @@ func (q *sliceBasedWaiterQueue) put(ws ...*waiter) {
 	}
 	q.waiters = append(q.waiters, ws...)
 	v2.TxnLockWaitersTotalHistogram.Observe(float64(len(q.waiters)))
+}
+
+func (q *sliceBasedWaiterQueue) remove(w *waiter) (bool, bool) {
+	q.Lock()
+	defer q.Unlock()
+
+	if q.beginChangeIdx != -1 {
+		panic("BUG: cannot remove waiter in changing waiter queue")
+	}
+
+	for i, v := range q.waiters {
+		if v != w {
+			continue
+		}
+		v.close("sliceBasedWaiterQueue remove", q.logger)
+		copy(q.waiters[i:], q.waiters[i+1:])
+		q.waiters[len(q.waiters)-1] = nil
+		q.waiters = q.waiters[:len(q.waiters)-1]
+		v2.TxnLockWaitersTotalHistogram.Observe(float64(len(q.waiters)))
+		return true, i == 0
+	}
+	return false, false
 }
 
 func (q *sliceBasedWaiterQueue) notifyAll(value notifyValue) {

